### PR TITLE
Adding pretrained ViT weights

### PR DESCRIPTION
**Pull request by sallysyw**
In #4594, we added ViT models' architecture to torchvision prototype. 
In this PR, we are going to add pretrained weights for ViT models :D

| Model      | Training Recipe | Training Command | Training Job ID | Epochs  | Nodes | Num of GPUs per node | Batch Size per GPU | Global batch\_size | Image Size | Representation Size | Original Paper Acc@1 | Classy Vision Acc@1 | ACC@1  | ACC@5  | Testing Job ID | Testing Command                                                                                                                                                                                                                                                                                 |
| ---------- | ---------- | ------ | --------------- | ------- | ----- | -------------------- | ------------------ | ------------------ | ---------- | ------------------- | -------------- | ------------------ | ------ | ------ | -------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| vit\_b\_16   | Close to DeiT | PYTHONPATH=$PYTHONPATH:`pwd` python -u run_with_submitit.py --timeout 3000 --ngpus 8 --nodes 8 --partition train --model vit_b_16 --batch-size 64 --epochs 300 --opt adamw --lr 0.003 --wd 0.3 --lr-scheduler cosineannealinglr --lr-warmup-method linear --lr-warmup-epochs 30 --lr-warmup-decay 0.033 --amp --label-smoothing 0.11 --mixup-alpha 0.2 --auto-augment ra --clip-grad-norm 1 --ra-sampler --cutmix-alpha 1.0 --model-ema |10266    | 268/300 | 8     | 8                    | 64                 | 4096               | 224        | None                | 77.91          | 78.98              | 81.072 | 95.318 | 13799  | PYTHONPATH=$PYTHONPATH:\`pwd\` python -u run\_with\_submitit.py --timeout 3000 --ngpus 1 --nodes 1 --partition train --model vit\_b\_16 --batch-size 1 --test-only --weights ViT\_B\_16\_Weights.ImageNet1K\_V1 |
| vit\_b\_32 | Close to DeiT  | PYTHONPATH=$PYTHONPATH:`pwd` python -u ~/workspace/scripts/run_with_submitit.py --timeout 3000 --ngpus 8 --nodes 2 --partition train --model vit_b_32 --batch-size 256 --epochs 300 --opt adamw --lr 0.003 --wd 0.3 --lr-scheduler cosineannealinglr --lr-warmup-method linear --lr-warmup-epochs 30 --lr-warmup-decay 0.033 --amp --label-smoothing 0.1 --mixup-alpha 0.2 --auto-augment imagenet --clip-grad-norm 1 --ra-sampler --cutmix-alpha 1.0 --model-ema |10265           |  291/300 | 2     | 8                    | 256                | 4096               | 224        | None                | 73.38          | 73.3               | 75.912 | 92.466 | 13796          | PYTHONPATH=$PYTHONPATH:\`pwd\` python -u run\_with\_submitit.py --timeout 3000 --ngpus 1 --nodes 1 --partition train --model vit\_b\_32 --batch-size 1 --test-only --weights ViT\_B\_32\_Weights.ImageNet1K\_V1 |
| vit\_l\_16 | TorchVision New Recipe  |PYTHONPATH=$PYTHONPATH:`pwd` python -u run_with_submitit.py --timeout 3000 --ngpus 8 --nodes 2 --model vit_l_16 --batch-size 64 --lr 0.5 --lr-scheduler cosineannealinglr --lr-warmup-epochs 5 --lr-warmup-method linear --auto-augment ta_wide --epochs 600 --random-erase 0.1 --label-smoothing 0.1 --mixup-alpha 0.2 --cutmix-alpha 1.0 --weight-decay 0.00002 --norm-weight-decay 0.0 --model-ema --val-resize-size 232 --clip-grad-norm 1 --ra-sampler  |12107/12349/12567 (due to multiple times of resuming)           | 378/600 | 2     | 8                    | 16                 | 1024               | 224        | None                | 76.53          | 76.57              | 79.662 | 94.638 | 13804          | PYTHONPATH=$PYTHONPATH:\`pwd\` python -u run\_with\_submitit.py --timeout 3000 --ngpus 1 --nodes 1 --partition train --model vit\_l\_16 --batch-size 1 --test-only --weights ViT\_L\_16\_Weights.ImageNet1K\_V1 |
| vit\_l\_32 | Close to DeiT  |PYTHONPATH=$PYTHONPATH:`pwd` python -u run_with_submitit.py --timeout 3000 --ngpus 8 --nodes 8 --partition train --model vit_l_32 --batch-size 64 --epochs 300 --opt adamw --lr 0.003 --wd 0.3 --lr-scheduler cosineannealinglr --lr-warmup-method linear --lr-warmup-epochs 30 --lr-warmup-decay 0.033 --amp --label-smoothing 0.11 --mixup-alpha 0.2 --auto-augment ra --clip-grad-norm 1 --ra-sampler --cutmix-alpha 1.0 --model-ema  |10430           | 253/300 | 8     | 8                    | 64                 | 4096               | 224        | None                | 71.16          | 73.49              | 76.972 | 93.07  | 13793          | PYTHONPATH=$PYTHONPATH:\`pwd\` python -u run\_with\_submitit.py --timeout 3000 --ngpus 1 --nodes 1 --partition train --model vit\_l\_32 --batch-size 1 --test-only --weights ViT\_L\_32\_Weights.ImageNet1K\_V1 |

cc @datumbox

### CODE REVIEW

File: torchvision/prototype/models/vision_transformer.py [hunk 1]
```diff
@@ -233,14 +235,40 @@ def forward(self, x: torch.Tensor):
...
+    "interpolation": InterpolationMode.BILINEAR,
+}
+
+
 class ViT_B_16_Weights(WeightsEnum):
-    # If a default model is added here the corresponding changes need to be done in vit_b_16
-    pass
+    ImageNet1K_V1 = Weights(
+        url="https://download.pytorch.org/models/vit_b_16-0413b9bf.pth",
+        transforms=partial(ImageNetEval, crop_size=224),
```
Comment (datumbox) [inline torchvision/prototype/models/vision_transformer.py:247]: Are you sure the weights are optimal without using resize_size optimizations?

**Review by datumbox — COMMENTED** [review]
Thanks @sallysyw. From #5086 I understand that some of the optimizations applied to all other models were not used here. Can you confirm that you tried them and were not beneficial? If you haven't used them, I would strongly recommend doing a few more runs prior merging this to confirm we pushed the accuracy as high as we can. The above tricks helped a multitude of models including ResNets, RegNets (unpublished, coming soon), MobileNetV3, EfficientNet and ResNeXt.

File: torchvision/prototype/models/vision_transformer.py (hunk 1 above)
Comment (sallysyw) [reply torchvision/prototype/models/vision_transformer.py:247]: Currently doing a batch of new training to see whether the accuracy can be further increased by adopting the new recipe.

Comment (sallysyw) [reply torchvision/prototype/models/vision_transformer.py:247]: Actually - i think the current results are publishable - they might not be the most optimal results, but they are good enough to beat the original paper's result and classy vision's benchmarking results. I admit that by adopting more fancy training techniques such as conv_stem or applying the tuned parameters in the new recipe can help us gain more points, but it does requires training more epochs (for the torchvision new training recipe it trains 600 epochs to converge). But for now, all the baseline results in other repo (such as google vit or deit) they always train 300 epochs (just as i did here), and i think we can keep aligned with them in imagenet1k_v1 (for reference). Probably in the future we can publish the new weights (trained with new recipe) and name it imagenet1k_v2. 
wdyt? @datumbox

File: torchvision/prototype/models/vision_transformer.py [hunk 2]
```diff
@@ -233,14 +235,40 @@ def forward(self, x: torch.Tensor):
...
+        meta={
+            **_COMMON_META,
+            "size": (224, 224),
```
[the remark this answers is not in the document]
Comment (sallysyw) [reply torchvision/prototype/models/vision_transformer.py:250]: I didn't do this for a reason - we might want to fine-tune the model on images with higher resolution and publish the results in the future, at that time the `size` would change to 384, that's why I didn't move it to `_COMMON_META` :D

File: torchvision/prototype/models/vision_transformer.py (hunk 1 above)
Comment (datumbox) [reply torchvision/prototype/models/vision_transformer.py:247]: @sallysyw I think we can do better. There are multiple pros in converging towards a generalizable recipe on which we can build upon rather than having custom recipes per model, especially if those custom recipes achieve lower accuracies as it is in this case. 

TorchVision's new recipe (published [here](https://pytorch.org/blog/how-to-train-state-of-the-art-models-using-torchvision-latest-primitives/)), when combined with some of the ViT specific optimizations you introduced (such as gradient clipping and the RA sampler) and some proposals from the [community](https://github.com/pytorch/vision/pull/5084#issuecomment-998369197), achieves much better results than the ones included in this PR. 

Here are some quick runs I did to confirm that the new recipe works well on ViT:
```
vit_b_16 (jobid 11379, epoch 450)
Acc@1 80.730 Acc@5 95.228

vit_l_32 (jobid 12096, epoch 378)
Acc@1 75.050 Acc@5 91.824

vit_l_16 (jobid 12107/12349/12567, epoch 378)
Acc@1 79.668 Acc@5 94.660
```

Given the above have 0.7-0.8 point higher accuracy, we should prefer them. Of course you can further optimize them by tuning the hyper-parameters but we should at least deploy the best available models. I've included the jobids so that you can find them easier on the internal infra but happy to chat offline to help.

Comment (sallysyw) [reply torchvision/prototype/models/vision_transformer.py:247]: Thanks for the info - I looked at the logs you provided and analyzed it using [this script](https://gist.github.com/kazhang/89525c80f2fe56caf25bb6779cf8cf33). And I found that the accuracy included in this PR are actually higher than the results suggested above - for example:

## 1. vit_b_16
### 1.1 The results trained with the new recipe (job-id 11379)
```
(base) sallysyw@ip-10-200-71-59:~/workspace$ python scripts/best_accuracy.py --job-ids 11379
Best acc@1      acc@5   epoch
78.764  93.334  588
Best EMA acc@1  acc@5   epoch
80.656  95.09   450
```

### 1.2 My previous result on vit_b_16 (job-id 10266)
```
(base) sallysyw@ip-10-200-71-59:~/workspace$ python scripts/best_accuracy.py --job-ids 10266
Best acc@1      acc@5   epoch
80.033  94.663  298
Best EMA acc@1  acc@5   epoch
81.082  95.32   268
```

## 2. vit_l_32
### 2.1 The results trained with the new recipe (job-id 12096)
```
(base) sallysyw@ip-10-200-71-59:~/workspace$ python scripts/best_accuracy.py --job-ids 12096
Best acc@1      acc@5   epoch
69.442  87.886  552
Best EMA acc@1  acc@5   epoch
75.05   91.824  378
```

### 1.2 My previous result on vit_b_32 (job-id 10265)
```
(base) sallysyw@ip-10-200-71-59:~/workspace$ python scripts/best_accuracy.py --job-ids 10265
Best acc@1      acc@5   epoch
75.646  92.164  296
Best EMA acc@1  acc@5   epoch
75.92   92.468  291
```

Looks like my best accuracy is always higher?
Is it because you are reporting the EMA accuracy but I'm reporting the non-EMA accuracy? If so, should I go with the EMA accuracy here?

**Comment by sallysyw** [comment]
Let me re-run the testing using the EMA weights rather than the original weights...

**Comment by sallysyw** [comment]
Here are the updated version of our best results (all tested with batch-size=1)
| Model      | Training Job ID | Epochs  | Nodes | Num of GPUs per node | Batch Size per GPU | Global batch\_size | Image Size | Representation Size | Original Acc@1 | ClassyVision Acc@1 | ACC@1  | ACC@5  | Testing Job ID | Testing Command                                                                                                                                                                                                                                                                                 |
| ---------- | --------------- | ------- | ----- | -------------------- | ------------------ | ------------------ | ---------- | ------------------- | -------------- | ------------------ | ------ | ------ | -------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| vit\_b\_16 | 10266           | 268/300 | 8     | 8                    | 64                 | 4096               | 224        | None                | 77.91          | 78.98              | 81.072 | 95.318 | 13799          | PYTHONPATH=$PYTHONPATH:\`pwd\` python -u ~/workspace/scripts/run\_with\_submitit.py --timeout 3000 --ngpus 1 --nodes 1 --partition train --model vit\_b\_16 --batch-size 1 --data-path /datasets01\_ontap/imagenet\_full\_size/061417/ --test-only --weights ViT\_B\_16\_Weights.ImageNet1K\_V1 |
| vit\_b\_32 | 10265           | 291/300 | 2     | 8                    | 256                | 4096               | 224        | None                | 73.38          | 73.3               | 75.912 | 92.466 | 13796          | PYTHONPATH=$PYTHONPATH:\`pwd\` python -u ~/workspace/scripts/run\_with\_submitit.py --timeout 3000 --ngpus 1 --nodes 1 --partition train --model vit\_b\_32 --batch-size 1 --data-path /datasets01\_ontap/imagenet\_full\_size/061417/ --test-only --weights ViT\_B\_32\_Weights.ImageNet1K\_V1 |
| vit\_l\_16 | 12349           | 378/600 | 8     | 8                    | 16                 | 1024               | 224        | None                | 76.53          | 76.57              | 79.662 | 94.638 | 13804          | PYTHONPATH=$PYTHONPATH:\`pwd\` python -u ~/workspace/scripts/run\_with\_submitit.py --timeout 3000 --ngpus 1 --nodes 1 --partition train --model vit\_l\_16 --batch-size 1 --data-path /datasets01\_ontap/imagenet\_full\_size/061417/ --test-only --weights ViT\_L\_16\_Weights.ImageNet1K\_V1 |
| vit\_l\_32 | 10430           | 253/300 | 8     | 8                    | 64                 | 4096               | 224        | None                | 71.16          | 73.49              | 76.972 | 93.07  | 13793          | PYTHONPATH=$PYTHONPATH:\`pwd\` python -u ~/workspace/scripts/run\_with\_submitit.py --timeout 3000 --ngpus 1 --nodes 1 --partition train --model vit\_l\_32 --batch-size 1 --data-path /datasets01\_ontap/imagenet\_full\_size/061417/ --test-only --weights ViT\_L\_32\_Weights.ImageNet1K\_V1 |

**Comment by datumbox** [comment]
@sallysyw Thanks for the clarifications and for making the necessary changes.

If I understand correctly, `vit_l_16` is produced using the new TorchVision recipe while `vit_b_16`, `vit_b_32` and `vit_l_32` use one that is closer to DeIT. If that's correct, we need to update the readme on the [references](https://github.com/pytorch/vision/tree/main/references/classification#vision-transformer) and your [table](https://github.com/pytorch/vision/pull/5085#issuecomment-1003247327) to reflect that.

For `vit_l_16` (jobids 12107/12349/12567) the exact config I used was:
```
 --ngpus 8 --nodes 2 --model vit_l_16 --batch-size 64 --lr 0.5 --lr-scheduler cosineannealinglr --lr-warmup-epochs 5 --lr-warmup-method linear --auto-augment ta_wide --epochs 600 --random-erase 0.1 --label-smoothing 0.1 --mixup-alpha 0.2 --cutmix-alpha 1.0 --weight-decay 0.00002 --norm-weight-decay 0.0 --model-ema --val-resize-size 232 --clip-grad-norm 1 --ra-sampler
```

We the above changes, we are getting close to being able to merge the PR. Please note that before doing so, we will need to finish some of the pending steps such as deploying the models on Manifold, adding them on the torchvision/models folders on our AWS infra. Let's finish these prior merging. I'll send you offline the guide that describes the model release process.

**Review by datumbox — APPROVED** [review]
LGTM, thanks a lot @sallysyw.

**Comment by xiaohu2015** [comment]
@sallysyw Is is a plan to add vit-tiny and vit-small (DeiT-Ti, DeiT-Small)
